script/refresh_sitemap

```diff
@@ -75,17 +75,17 @@ OBJECT_TYPES = [
 ].freeze
 
 STATIC_PAGES = [
-  ["home page",    "observer/list_rss_logs", "hourly"],
-  ["introduction", "observer/intro",         "weekly"],
-  ["how to use",   "observer/how_to_use",    "weekly"],
-  ["how to help",  "observer/how_to_help",   "weekly"]
+  ["home page",    "activity_logs/index", "hourly"],
+  ["introduction", "info/intro",          "weekly"],
+  ["how to use",   "info/how_to_use",     "weekly"],
+  ["how to help",  "info/how_to_help",    "weekly"]
 ].freeze
 
 # These are aliases for pages allowed above.
 OTHER_ALLOWED_PATHS = [
   # "/observer/index", (would allow too many indexes we don't want to allow)
-  "/observer/lookup_observation",
-  "/observer/lookup_name",
+  "/lookups/lookup_observation",
+  "/lookups/lookup_name",
```

Comment:
I'd guess that a sitemap should be an index of active routes, and the 301's should not be in it. Worth researching.

Reply:
Hm. well, discoverlife.org is offline. Who else

I have to say i disagree about permanent URLs — this amounts to an obligation to other websites never to refactor our site, which means ever-increasing technical debt and garble for us. 301's should be sufficient — they get legacy links where they need to go, and in my opinion, even those should have an expiration date. We have to adjust for their changing organizations, why not the other way around?

Certainly URLs are not considered permanent in other organizations. If a site gives a 301, it's considered a clients' responsibility to update their links, in my experience.

Reply:
For sitemaps, this is the kind of info i've been working with:

> These (XML) sitemaps give search engines a website’s URLs and offer data a complete map of all pages on a site. This helps search engines prioritize pages that they will crawl.
> 
> There is information within the sitemap that shows page change frequency on one URL versus others on that website, but it is unlikely that this has any effect on rankings.
> 
> An XML sitemap is very useful for large websites that might otherwise take a long time for a spider to crawl through the site and can help search engines just simply know what pages exist. Without this sitemap, they would not be able to build the full menu of available pages until they completed a crawl of the whole site.

I don't think update frequency has much effect on bot behavior; they're trying to crawl the whole thing. On the other hand, seems like we could keep things off the sitemap that we don't want crawled.

Reply:
I couldn't disagree more strongly with you, Nimmo!  One, it doesn't prevent us from refactoring in the slightest.  Two, sites that change their canonical URLs are doing a huge disservice to the larger web community.  All it means is separating the public view from the internals of the site.  And that's just plain good programming practice.  It's convenient at first for the URLs to match internal labels, sure.  But to *rely* on that, as you say, forces us to either never refactor or to change the public interface willy nilly with every whim of the internal coding structure.

Perhaps we're misunderstanding each other.  I am only advocating for a public route /observer/lookup_name to stick around.  You can route it to whatever internal action you wish.  If you don't like the parameter format, we can make a wrapper that converts it.  You are free to refactor the controller and name it whatever you wish.  I don't see how keeping /observer/lookup_name puts any constraints on your refactor whatsoever!

Reply:
OK - Currently the `observer/lookup_name` routes are still there, but they are 301'd. 
You're saying: keep them there and **active**, not 301?

Either way you're right, this is not a big deal to maintain.

For background, my sense — coming from the bot/API/EOL/discoverlife.org situations — is that our code maintains a lot of things out of consideration for phantom "participants" who may be... hoodlum bots not even interested in mycology, or the original participants may not even be there at all, or that our code intended to provide for them may be broken and we're not aware. Maybe we can start implementing stronger customs of maintaining contact info for all end users...? There can't be _that_ many academic, institutional or even solo end users of MO... and my suspicion is that they're all capable of writing and receiving an email. It's not like we're providing translations of banned Chinese political texts, and the users need the anonymity... or do they?

Reply:
and by solo end users i mean, end users of custom-tailored routes and functions, not of the site generally.

Reply:
I guess what i'm saying is...\ i have a hard time mustering mental space for anyone or any institution who expects to mooch data off MO without so much as a courtesy email introducing themselves, and providing contact information. Heck, I introduced myself to you 11 years ago and I didn't even want anything! I'm persuadable though, if you can think of a situation where this kind of lurker mooching is valuable and unavoidable.

Reply:
Good conversation!  And I think your attitude is probably the healthier one in general.  I do think of it a little differently, though.  Not in terms of what people expect of us, but of what I've (we've) promised other people.  There aren't many things, so I don't think this is as big as you are worried.

1. image URLs -- https://mushroomobserver.org/images/$size/$id.jpg -- I've been trying to get people to use this URL for ages.  From the beginning, really, but that's literally where they were at the beginning, back when we actually stored the images locally on the webserver(!!)

2. observations at https://mushroomobserver.org/$id

3. content we've "published" -- By this, I mean basic observation data, including notes and comments.  This includes markup, in which you can find these lookup_xxx URLs.  We decided early on not to strip HTML / markup out of the notes.  That means these have been available for many years.  Many, many people have contacted us asking about downloading data from MO.  Foolishly perhaps, I have never bothered to keep track of them.  My position was, okay, fine, the easiest way to manage access is simply to give it away without any restrictions.  That meant structuring the markup so that it would be guaranteed to be stable.  So I created the _blah_ mechanism for creating internal links, made them point to this set of silly lookup_blah actions with the express (to myself at least) expectation that they be guaranteed to remain stable.  But nothing else.  We do allow users to enter their own links and HTML and stuff, but that's all implicitly at their own risk.  I make no guarantees, obviously, for that sort of thing.  I designed the lookup mechanism the way I did so that it did not rely on name ids, because the meaning of a name id could potentially change.  Instead I made it go by the scientific name.  I figured that was as stable and simple and transparent as possible.

That's about it.  Maybe there's something else somewhere, but I can't think of anything else right now.  I have never encouraged third parties to link directly to name pages, or project pages or anything else.  Just these three things.  The API is there for anything else.  And of course, that's a whole other can of worms.

So if we can just ensure these three canonical URLS -- images, observations, lookups -- I will be happy.  Redirects are okay, I guess, if we must, but rails makes it so easy to route them directly, so why not?  /1234 doesn't redirect to /observations/1234, it routes straight there.  I wish I could do that with the images, too, but I've never figured out how.  The problem with the redirect is people get the 301 or whatever the code is, and think "darn MO, they've changed it, I thought that was supposed to be stable, oh well, I guess I'll change it".  But that's wrong!  I do *not* guarantee the new route, only the old route.  So the redirect fools people into switching to an unstable route.  Ugh.  That's caused so many people problems with the images.  Including people whom I've told very explicitly in no uncertain terms not to do precisely that!  Arg.  So, yes, I would prefer rerouting, not redirecting, for that reason! :)

Reply:
Got it. Lookup routes shall endure!

For observations and images, I'll explore if there's a way to redirect and send a different HTTP code. `300` means "multiple choice", i.e. we're rerouting you, but both urls are valid. Maybe that's the one we should try — but TBD if Rails allows customizing the response code. (Happy to leave that alone, also... Is this an interesting thing to try?)

Reply:
Regarding redirecting images and observations, yes, it is possible. Idk if this is what you intend, but it does redirect and the status says "both routes valid".
```ruby
get ':id', to: redirect('/observations/%{id'}', status: 300)
```

Reply:
Interesting.  I think I'd like /$id to reroute not redirect, anyway... yes, the way it works right now.  And the image URLs are handled by nginx so they never make it to the rails router.  But maybe there's a way to tell nginx to use 300 instead of 301 when "rewriting" a request.  Thanks!